052-cruise-control-api-users.md

```diff
@@ -0,0 +1,269 @@
+userOne: passwordOne, USER
+userTwo: passwordTwo, VIEWER
+```
+
+Then the developer can use this file to create a secret using the following command:
+```
+kubectl create secret generic cruise-control-api-users-secret  --from-file=key=cruise-control-auth.txt
+```
+
+The command would create a secret:
+```yaml=
+apiVersion: v1
+kind: Secret
+metadata:
+  name: cruise-control-api-users-secret
+type: Opaque
+data:
+  key: dXNlck9uZTogcGFzc3dvcmRPbmUsIFVTRVIKdXNlclR3bzogcGFzc3dvcmRUd28sIFZJRVdFUgo=
+```
+
+The developer could then reference the secret in a `spec.cruiseControl.apiUsers` section of the `Kafka` resource.
+
+The schema would look like this:
+
+```yaml=
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  ...
+spec:
+  cruiseControl:
+    apiUsers:
+      type: hashloginservice (1)
+      valueFrom: (2)
+        secretKeyRef:
+          name: cruise-control-api-users-secret
+          key: key
+     ...
+```
+(1) A `type` field is added here to describe the format of the data and for configuration flexibility in the future. Thsi gives us the option to add different data format types in the future. 
+In this example we use Jetty's HashLoginService format since that is the format which Cruise Control uses for its API user configuration.
+
+(2) The `valueFrom` construct allows us to add more sources in the future if needed. 
+This is also a pattern used in other Strimzi APIs already, for example Strimzi's logging configuration, password configuration in KafkaUser resources, metrics configuration, and more.
+
+Strimzi would decode and use the contents of this secret to populate the Cruise Control API auth credentials file which would be mounted and used by the Cruise Control pod.
+
+Although this would simplify the overall design it would also shift a lot of the maintanence responsibility to the developer.
+The developer would need to create a valid Cruise Control API user configuration and encode it into a secret.
+That being said, this is still a reasonable amount of responsibility to leave to the developer.
+This is because accessing the Cruise Control API directly is not the primary way of interacting with Cruise Control for a Strimzi-managed Kafka cluster and is expected to be used only by advanced developers in special use-cases.
+However, with proper supporting documentation, this new feature will give developers a greater degree of control and assistance with special use-cases.
+
+### Risks
+
+Direct calls to the Cruise Control API like the write operations enabled by the `ADMIN` role could interfere with Strimzi Operator operations and cause damage to a Strimzi-managed Kafka cluster.
+
+Supporting developers to create Cruise Control API users with non-`ADMIN` roles like `USER` and `VIEWER` roles would mitigate this risk completely.
+This is because the `USER` and `VIEWER` roles are read-only roles and would not conflict with Strimzi cluster operations.
+So this proposal is for enabling the `USER` and `VIEWER` roles only.
```

Comment:
I'd like for the proposal to define:

* What will happen if the user specifies a file with the `ADMIN` role? 
* What will happen if the user specifies a corrupt file (not in the expected format)?

Reply:
Yeah. How these error conditions are reflected in the Kafka CR status?

Reply:
@fvaleri They wouldn't be for now, for a feature like this it would not be worth the extra effort and code complexity. If needed, we could always add it at a later date